src/sparseml/optim/helpers.py

```diff
@@ -438,3 +444,86 @@ def _maybe_parse_number(val: str) -> Union[str, float, int]:
             return float(val)
         except Exception:
             return val
+
+
+def validate_metadata(metadata: dict, yaml_str: str) -> dict:
+    """
+    Merge the metadata carried over from the recipe (`yaml_str`) with
+    the new, incoming metadata ('metadata').
+
+    If yaml_str is a staged recipe, this function returns a multi-key
+        dict where:
+        -keys are stage names
+        -values are metadata for respective stage name
+    If yaml_str is a normal recipe, this function returns a single-key
+        dict where:
+        -key is '__metadata__'
+        -value is metadata for the recipe
+
+    If there exists a key in new metadata which also exists in previous metadata,
+    it will overwrite the old metadata. However, it is prohibitive to pass novel keys.
+
+    :param metadata: New metadata
+    :param yaml_str: String representation of the recipe YAML file,
+        (may contains previous metadata)
+    :return: Validated metadata
+    """
+
+    default_metadata_key = "__metadata__"
+    _, container = evaluate_recipe_yaml_str_equations(yaml_str)
```

Comment:
to get the container, using `load_recipe_yaml_str_no_classes` should be sufficient, no need to fuss with eval